send-request.js

```diff
@@ -0,0 +1,109 @@
+import http from "http";
+import { uuid } from "mu";
+
+const MAX_RETRIES = process.env.MAX_RETRIES || 3;
+const RETRY_TIMEOUT = process.env.RETRY_TIMEOUT || 250;
+
+function formatChangesetBody(changeSets, options) {
+  if (options.resourceFormat == "v0.0.1") {
+    return JSON.stringify(
+      changeSets.map((change) => {
+        return {
+          inserts: change.insert,
+          deletes: change.delete,
+        };
+      })
+    );
+  }
+  if (options.resourceFormat == "v0.0.0-genesis") {
+    // [{delta: {inserts, deletes}]
+    const newOptions = Object.assign({}, options, { resourceFormat: "v0.0.1" });
+    const newFormat = JSON.parse(formatChangesetBody(changeSets, newOptions));
+    return JSON.stringify({
+      // graph: Not available
+      delta: {
+        inserts: newFormat
+          .flatMap(({ inserts }) => inserts)
+          .map(({ subject, predicate, object }) => ({
+            s: subject.value,
+            p: predicate.value,
+            o: object.value,
+          })),
+        deletes: newFormat
+          .flatMap(({ deletes }) => deletes)
+          .map(({ subject, predicate, object }) => ({
+            s: subject.value,
+            p: predicate.value,
+            o: object.value,
+          })),
+      },
+    });
+  } else {
+    throw `Unknown resource format ${options.resourceFormat}`;
+  }
+}
+
+export async function sendRequest(
+  entry,
+  changeSets,
+  muCallIdTrail,
+  muSessionId,
+  extraHeaders = {},
+  retries = MAX_RETRIES
+) {
+  // construct the requestObject
+  const method = entry.callback.method;
+  const url = entry.callback.url;
+  const headers = {
+    ...extraHeaders,
+    "Content-Type": "application/json",
+    "MU-AUTH-ALLOWED-GROUPS": changeSets[0].allowedGroups,
+    "mu-call-id-trail": muCallIdTrail,
+    "mu-call-id": uuid(),
+    "mu-session-id": muSessionId,
+  };
+
+  let body;
+  if (entry.options && entry.options.resourceFormat) {
+    // we should send contents
+    body = formatChangesetBody(changeSets, entry.options);
+  }
+  if (process.env["DEBUG_DELTA_SEND"])
+    console.log(`Executing send ${method} to ${url}`);
+  try {
+    const keepAliveAgent = new http.Agent({
+      keepAlive: true,
+    });
+    const response = await fetch(url, {
+      method,
+      headers,
+      body,
+      agent: keepAliveAgent,
+    });
+    if (!response.ok) {
+      console.log(
+        `Call to ${method} ${url} likely failed. Received status ${response.status}.`
+      );
+      throw new Error(`failed to send request, status: ${response.status}`);
```

Comment:
this throw is new, previously, you only got the 'likely failed' message. I believe a not ok status really means it failed and we should retry if we do retries, hence the throw.

Reply:
Is this throw caught somewhere?  If not, does this bring down the delta-notifier?

Reply:
I personally find the logging a bit too noisy now that we're throwing here, but not an issue with the throw itself. (perhaps the `console.log(error)` line should be removed?

Reply:
hmm yes maybe we should only log the full error if the final retry fails?

Reply:
I also think it doesn't make sense to retry some error codes. E.g. most 4xx codes will not be resolved by retrying.

Reply:
that's fair, i'll just do status >= 500 (and !status in case the request fails before a status code)?